### PR TITLE
bazel: upgrade `rules_go`

### LINKED ISSUE #69534: testing: github auto-filer does not surface race condition details properly
See for example #69522

image

The top of the captured stack trace does not properly capture the start of the race condition.

### LINKED ISSUE #85650: ci: test failures caused by panics are not clear
Here is a Github CI run that shows a test failure on `TestConnector_OpenTenantConnWithToken/error_during_auth` caused by a panic: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests/5978643?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true
image

Here is the Bazel CI run for the same PR, which is missing this info: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/5978667?showRootCauses=false&expandBuildChangesSection=true&expandBuildDeploymentsSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true&logFilter=debug
image

Jira issue: CRDB-18358
Epic CRDB-8349

### LINKED ISSUE #87186: ci/bazel: In CI, test stderr should be collected in test.xml as well as test.log
This can manifest as `No pass/skip/fail event found for test` as in [internal discussion](https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1661891000169659?thread_ts=1661885577.894459&cid=CJ0H8Q97C).

Epic CRDB-15060

Jira issue: CRDB-19198

### PR DESCRIPTION
**Pull request by rickystewart**
Pull in cockroachdb/rules_go#7.

Closes #69534.
Closes #85650.
Closes #87186.

Release justification: Non-production code changes
Release note: None

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -18,12 +18,12 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
...
-    sha256 = "12f619147ab74d6d50741652ba1cdd48707d80ffd13a2cc79fb630265ff03de5",
-    strip_prefix = "cockroachdb-rules_go-cb54c66",
+    sha256 = "0f95eeba78dc5eeea37673039ac9db00eb25d1a2c65a7c44779678ad460ced95",
+    strip_prefix = "cockroachdb-rules_go-4e5db62",
     urls = [
-        # cockroachdb/rules_go as of cb54c66bc3e23a9533a2d04459f6262a6abb8fad
+        # cockroachdb/rules_go as of 4e5db6270b33b19f1ebcafee5dd97f54f990afdb
         # (upstream release-0.34 plus a few patches).
-        "https://storage.googleapis.com/public-bazel-artifacts/bazel/cockroachdb-rules_go-v0.27.0-164-gcb54c66.tar.gz",
+        "https://storage.googleapis.com/public-bazel-artifacts/bazel/cockroachdb-rules_go-v0.27.0-165-g4e5db62.tar.gz",
```
Comment (healthy-pod) [inline WORKSPACE:26]: What is `v0.27.0` referring to? If I understand correctly it is the [most recent synced tag from upstream](https://github.com/cockroachdb/rules_go/tags) but we are using a patched `v0.34.0` so should it be `v0.34.0`? I also see that we once changed that to `v0.29.0` then changed it back to `0.27.0` so I am sure I am missing something.

Comment (rickystewart) [reply WORKSPACE:26]: It's not really meaningful, it's just the name GitHub gave the tarball. https://github.com/cockroachdb/rules_go/tarball/release-0.34

**Review by healthy-pod — APPROVED** [review]

**Comment by rickystewart** [comment]
bors r=healthy-pod

**Comment by rickystewart** [comment]
blathers backport 22.1

**Comment by rickystewart** [comment]
blathers backport 22.2

**Pull request merged**